tests/microos/network_bonding.pm

```diff
@@ -11,45 +11,149 @@ use strict;
 use warnings;
 use testapi;
 use power_action_utils "power_action";
+use serial_terminal qw(select_serial_terminal);
+
+sub check_connectivity {
+    my ($bond_name) = @_;
+    my $ping_host = "conncheck.opensuse.org";
+    my $ping_command = "ping -c1 -I $bond_name $ping_host";
+    assert_script_run $ping_command;
+}
+
+sub get_nics {
+    my ($bond_name) = @_;
+    my @devices;
+
+    foreach my $line (split('\n', script_output('nmcli -g DEVICE conn show'))) {
+        # Skip empty lines
+        next if $line =~ /^\s*$/;
+
+        # Skip the loopback device and the bond interface
+        next if $line eq 'lo';
+        next if $line eq $bond_name;
+
+        # Add valid device names to the array
+        push @devices, $line;
+    }
+
+    return @devices;
+}
+
+sub delete_existing_connections {
+    my $output = script_output('nmcli -g DEVICE,UUID conn show');
+    my %seen_uuids;
+
+    foreach my $line (split "\n", $output) {
+        next if $line =~ /^\s*$/;
+
+        my ($device, $uuid) = split /:/, $line;
+        next if defined $device && $device eq 'lo';
+        next if exists $seen_uuids{$uuid};
+
+        $seen_uuids{$uuid} = 1;
+        script_run "nmcli con delete uuid '$uuid'";
+    }
+}
+
+sub create_bond {
+    my ($bond_name, $bond_mode, $miimon) = @_;
+    assert_script_run "nmcli con add type bond ifname $bond_name con-name $bond_name bond.options \"mode=$bond_mode, miimon=$miimon\"";
+    assert_script_run "nmcli connection modify $bond_name connection.autoconnect-slaves 1";
+}
+
+sub add_devices_to_bond {
+    my ($bond_name, @devices) = @_;
+    foreach my $device (@devices) {
+        assert_script_run "nmcli con add type ethernet ifname $device master $bond_name";
+    }
+}
 
 sub test_failover {
-    my $device = shift;
-    # disable one child eth, other should keep bond0 alive
+    my ($bond_mode, $bond_name, $device, $description, $nics_ref) = @_;
+    my $delay = 5;
+    my @nics_status = map { [$_, $_ eq $device ? 0 : 1] } @$nics_ref;
+
+    record_info("Testing Failover for Mode: $bond_mode", "NIC: $device");
+
     assert_script_run "ip link set dev $device down";
+    sleep $delay;
     script_run 'ip a';
```

Comment:
It would be better to use one of the `_retry` functions for this one e.g `validate_script_output_retry`